go/api/config.go

```diff
@@ -0,0 +1,320 @@
+	} else {
+		request.TlsMode = protobuf.TlsMode_NoTls
+	}
+
+	if config.credentials != nil {
+		if err := config.credentials.validate(); err != nil {
+			return nil, err
+		}
+
+		authInfo := protobuf.AuthenticationInfo{}
+		if config.credentials.Username != nil {
+			authInfo.Username = *config.credentials.Username
+		}
+		authInfo.Password = *config.credentials.Password
+		request.AuthenticationInfo = &authInfo
+	}
+
+	request.ReadFrom = mapReadFrom(config.readFrom)
+	if config.requestTimeout != nil {
+		request.RequestTimeout = *config.requestTimeout
+	}
+
+	return &request, nil
+}
+
+func mapReadFrom(readFrom ReadFrom) protobuf.ReadFrom {
+	if readFrom == PreferReplica {
+		return protobuf.ReadFrom_PreferReplica
+	}
+
+	return protobuf.ReadFrom_Primary
+}
+
+// BackoffStrategy represents the strategy used to determine how and when to reconnect, in case of connection failures. The
+// time between attempts grows exponentially, to the formula rand(0 ... factor * (exponentBase ^ N)), where N is the number of
+// failed attempts.
+//
+// Once the maximum value is reached, that will remain the time between retry attempts until a reconnect attempt is successful.
+// The client will attempt to reconnect indefinitely.
+type BackoffStrategy struct {
+	// Required: number of retry attempts that the client should perform when disconnected from the server, where the time
+	// between retries increases. Once the retries have reached the maximum value, the time between retries will remain
+	// constant until a reconnect attempt is successful.
+	NumOfRetries *uint32
+	// Required: the multiplier that will be applied to the waiting time between each retry.
+	Factor *uint32
+	// Required: the exponent base configured for the strategy.
+	ExponentBase *uint32
+}
+
+func (strategy *BackoffStrategy) validate() error {
+	if strategy.NumOfRetries == nil {
+		return &RedisError{
+			"BackoffStrategy.NumOfRetries evaluated to nil. NumOfRetries must be non-nil to use a BackoffStrategy.",
+		}
+	}
+
+	if strategy.Factor == nil {
+		return &RedisError{
+			"BackoffStrategy.Factor evaluated to nil. Factor must be non-nil to use a BackoffStrategy.",
+		}
+	}
+
+	if strategy.ExponentBase == nil {
+		return &RedisError{
+			"BackoffStrategy.ExponentBase evaluated to nil. ExponentBase must be non-nil to use a BackoffStrategy.",
+		}
+	}
+
+	return nil
+}
+
+// RedisClientConfiguration represents the configuration settings for a Standalone Redis client. baseClientConfiguration is an
+// embedded struct that contains shared settings for standalone and cluster clients.
+type RedisClientConfiguration struct {
+	baseClientConfiguration
+	reconnectStrategy *BackoffStrategy
+	databaseId        *uint32
+}
+
+// NewRedisClientConfiguration returns a [RedisClientConfiguration] with default configuration settings. For further
+// configuration, use the [RedisClientConfiguration] With* methods.
+func NewRedisClientConfiguration() *RedisClientConfiguration {
```

Comment:
Is there a way in go to make contstructor private?

Reply:
Yep, in Go if the symbol starts with lower case (eg newRedisClientConfiguration) it is private. In this case, the only point of the constructor is for users, so it probably doesn't make sense to make it private. We could get rid of the constructor, but I wonder if its more idiomatic to provide it when the struct has a builder pattern. Without it the code looks a bit strange if we're using the builder pattern, it would look like this:
```
config := RedisClientConfiguration{}.
    WithUseTLS(true)
```